beacon_node/eth2_libp2p/src/peer_manager/mod.rs

```diff
@@ -940,19 +947,31 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
         let connected_peer_count = self.network_globals.connected_peers();
         if connected_peer_count > self.target_peers {
             //remove excess peers with the worst scores, but keep subnet peers
-            for (peer_id, _) in self
+            //must also ensure that the outbound-only peer count does not go below the minimum threshold
+            let mut peers_removed = 0;
+            let mut n_outbound_removed = 0;
+            for (peer_id, info) in self
                 .network_globals
                 .peers
                 .read()
                 .worst_connected_peers()
                 .iter()
                 .filter(|(_, info)| !info.has_future_duty())
-                .take(connected_peer_count - self.target_peers)
-                //we only need to disconnect peers with healthy scores, since the others got already
-                //disconnected in update_peer_scores
-                .filter(|(_, info)| info.score_state() == ScoreState::Healthy)
             {
-                disconnecting_peers.push(**peer_id);
+                if peers_removed == connected_peer_count - self.target_peers {
+                    break;
+                }
+                if info.is_outbound_only() {
+                    if n_outbound_removed < outbound_only_peer_count - min_outbound_only_target {
+                        n_outbound_removed += 1;
+                    } else {
+                        continue;
+                    }
+                }
+                peers_removed += 1;
+                if info.score_state() == ScoreState::Healthy {
+                    disconnecting_peers.push(**peer_id);
```

Comment:
Seems here that we only disconnect peers that are in the healthy state (this is the correct logic) but you are incrementing the removed count outside of this statement, similarly with `n_outbound_remove`. The result looks like we might not actually disconnect as many nodes as we think we are.

Also, you could just use `disconnecting_peers.len()` rather than tracking the separate variable `peers_removed` to enforce consistency.

Reply:
I think this behavior is meant to mimic the logic of the original code, which took ```connected_peer_count - self.target_peers```  peers before filtering out the ones that are not healthy. This is also why I used ```peers_removed``` instead of ```disconnecting_peers.len()```

Reply:
Oh yeah. Looks to me like its a bug in the old impl. I think the old logic is probably wrong here. Nice catch

Reply:
I actually think the old impl might be fine? Since as the comments mention, update_peer_scores will have disconnected the unhealthy ones already, so it's fine to not add them to the disconnecting_peers list since it's redundant. Or I might be missing something. It def tripped me up looking at the ordering of the filters

Reply:
Ah yes. Sorry, its been a while since I was in here and this has more complex logic than I recall. 

Yeah, looks like the old logic is correct, for peers that have transitioned into a disconnected or worse state we should try and disconnect from them. They could appear here if they are still connected and we want to avoid sending a double disconnection message. 

Your logic here looks correct, sorry I only did a quick skim of the PR :p.

Reply:
No worries! You guys got a lot of work goin on, just glad I wasn't totally off with my implementation